examples/src/main/java/com/squareup/moshi/recipes/Unwrap.java

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import com.squareup.moshi.recipes.Unwrap.EnvelopeJsonAdapter.Enveloped;
+import com.squareup.moshi.recipes.models.Card;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class Unwrap {
+  @Enveloped private static Object unused;
+
+  public static void main(String[] args) throws Exception {
+    String json = ""
+        + "{\"data\":"
+        + "  {\n"
+        + "    \"rank\": \"4\",\n"
+        + "    \"suit\": \"CLUBS\"\n"
+        + "  }"
+        + "}";
+    Annotation annotation = Unwrap.class.getDeclaredField("unused").getAnnotations()[0];
+    Moshi moshi = new Moshi.Builder().add(EnvelopeJsonAdapter.FACTORY).build();
+    JsonAdapter<Card> adapter =
+        moshi.adapter(Card.class, Collections.<Annotation>singleton(annotation));
+    Card out = adapter.fromJson(json);
+    System.out.println(out);
+  }
+
+  public static final class EnvelopeJsonAdapter extends JsonAdapter<Object> {
+    public static final JsonAdapter.Factory FACTORY = new Factory() {
+      @Override
+      public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+        Set<? extends Annotation> delegateAnnotations =
+            delegateAnnotations(annotations, Enveloped.class);
+        if (delegateAnnotations == null) {
+          return null;
+        }
+        Type envelope =
+            Types.newParameterizedTypeWithOwner(EnvelopeJsonAdapter.class, Envelope.class, type);
+        JsonAdapter<Envelope<?>> delegate = moshi.nextAdapter(this, envelope, delegateAnnotations);
+        return new EnvelopeJsonAdapter(delegate);
+      }
+    };
+
+    @Retention(RUNTIME) @JsonQualifier @interface Enveloped {
+    }
+
+    private static final class Envelope<T> {
+      final T data;
+
+      Envelope(T data) {
+        this.data = data;
+      }
+    }
+
+    private final JsonAdapter<Envelope<?>> delegate;
+
+    EnvelopeJsonAdapter(JsonAdapter<Envelope<?>> delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override public Object fromJson(JsonReader reader) throws IOException {
+      return delegate.fromJson(reader).data;
+    }
+
+    @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+      delegate.toJson(new Envelope<>(value));
+    }
+  }
+
+  public static Set<? extends Annotation> delegateAnnotations(Set<? extends Annotation> annotations,
```

Comment:
Yeah, the weirdness is that this is both a `hasJsonQualifier` and `nextAnnotations`, as you said.
At the same time, this seems to be the best way to accomplish this.
I'll send a PR and discuss there.